WalletWasabi/Bases/PeriodicRunner.cs

```diff
@@ -1,132 +1,105 @@
 		protected override async Task ExecuteAsync(CancellationToken stoppingToken)
 		{
 			while (!stoppingToken.IsCancellationRequested)
 			{
+				_tcs = new TaskCompletionSource<bool>();
+
 				try
 				{
+					// Do user action.
 					await ActionAsync(stoppingToken).ConfigureAwait(false);
-					LogAndResetLastExceptionIfNotNull();
+
+					// Log previous exception if any.
+					LogLastException(ExceptionTracker.LastException);
 				}
 				catch (Exception ex) when (ex is OperationCanceledException || ex is TaskCanceledException || ex is TimeoutException)
 				{
 					Logger.LogTrace(ex);
 				}
 				catch (Exception ex)
 				{
-					// Only log one type of exception once.
-					if (LastException is null // If the exception never came.
-						|| ex.GetType() != LastException.GetType() // Or the exception have different type from previous exception.
-						|| ex.Message != LastException.Message) // Or the exception have different message from previous exception.
+					// Exception encountered, process it.
+					LogLastException(ExceptionTracker.Process(ex));
+
+					Logger.LogError(ExceptionTracker.LastException!.Exception);
+				}
+
+				// Wait for the next round.
+				try
+				{
+					using var cts = CancellationTokenSource.CreateLinkedTokenSource(stoppingToken);
+					var linkedTcs = _tcs; // Copy reference so it cannot change.
+
+					if (linkedTcs.Task == await Task.WhenAny(linkedTcs.Task, Task.Delay(Period, cts.Token)).ConfigureAwait(false))
 					{
-						// Then log and reset the last exception if another one came before.
-						LogAndResetLastExceptionIfNotNull();
-						// Set new exception and log it.
-						LastException = ex;
-						LastExceptionFirstAppeared = DateTimeOffset.UtcNow;
-						LastExceptionCount = 1;
-						Logger.LogError(ex);
+						cts.Cancel(); // Ensure that the Task.Delay task is cleaned up.
 					}
 					else
 					{
-						// Increment the exception counter.
-						LastExceptionCount++;
+						linkedTcs.TrySetCanceled(); // Ensure that the tcs.Task is cleaned up.
 					}
 				}
-				finally
+				catch (TaskCanceledException ex)
 				{
-					try
-					{
-						CancellationToken? triggeringToken = null;
-						lock (TriggerLock)
-						{
-							triggeringToken = TriggeringCts?.Token;
-						}
-						if (triggeringToken is { })
-						{
-							using var linked = CancellationTokenSource.CreateLinkedTokenSource(stoppingToken, triggeringToken.Value);
-							await Task.Delay(Period, linked.Token).ConfigureAwait(false);
-						}
-					}
-					catch (TaskCanceledException ex)
-					{
-						Logger.LogTrace(ex);
-					}
-					finally
-					{
-						lock (TriggerLock)
-						{
-							if (TriggeringCts?.IsCancellationRequested is true)
-							{
-								TriggeringCts?.Dispose();
-								TriggeringCts = new CancellationTokenSource();
-							}
-						}
-					}
+					Logger.LogTrace(ex);
 				}
 			}
 		}
 
-		private void LogAndResetLastExceptionIfNotNull()
+		private void LogLastException(ExceptionInfo? info)
```

Comment:
I would make this part of the `ExceptionTracker` class because in that way all the possible `null` can be removed.